CTMU documentation update (multi-cluster and multi-version caching support) (#57325)
gravitational/teleport: docs/pages/upgrading/client-tools-managed-updates.mdx

folder and re-execute them if found.

**Validation**: Downloaded packages are verified with a hash sum to ensure integrity.
Package downloads are directed to the `cdn.teleport.dev` endpoint and depend on the operating system,
platform, and edition. The edition must be identified by the original client tools binary.
The URL pattern is as follows:
- `https://cdn.teleport.dev/teleport-{ent-}vX.Y.Z-{linux,darwin,windows}-{amd64,arm64,arm,386}-{fips-}bin.{tar.gz,pkg,zip}`
- `https://cdn.teleport.dev/teleport-{ent-}vX.Y.Z-{linux,darwin,windows}-{amd64,arm64,arm,386}-{fips-}bin.{tar.gz,pkg,zip}.sha256`
The base URL of the CDN can be overridden using the `TELEPORT_CDN_BASE_URL` environment variable.
Allows mirroring the CDN in a private network or using custom builds.

**Concurrency**: Tools use a locking mechanism to enable smooth operation during updates.
Only one process can acquire the lock to update client tools, while other processes wait for the lock to be released.
If the first process cancels the update, the next process in line will initiate the update.

### Multi-cluster support and multi-version caching

> Client tools managed updates support multi-cluster environments. When you log in to different clusters that use

Reviewer: Can we say "since version X.Y.Z"? Or maybe only in the backport?

Author: I going add this to each backport separately

If multiple clusters require the same version, a single shared copy is used. Once the update process is complete,
the relative path to the client tools is recorded in the configuration file.

```json
{
  "version": "v1",
  "configs": {
    "proxy1.example.com": {"version": "18.1.0","disabled": false}
  },
  "tools": [
    {
      "version": "18.1.0",
      "path": {
        "tctl": "1921b970-807f-4d36-a769-fbda149d8970-update-pkg-v2/tctl-18.1.0.pkg/Payload/tctl.app/Contents/MacOS/tctl",
        "tsh": "1921b970-807f-4d36-a769-fbda149d8970-update-pkg-v2/tsh-18.1.0.pkg/Payload/tsh.app/Contents/MacOS/tsh"
      }
    }
  ],
  "max_tools": 3
}
```

After each `tctl` or `tsh` command is re-executed using a specific version from the tools list,
the list is reordered to move the most recently used version to the top. If the number of cached
versions exceeds the allowed limit, the least recently used version (at the bottom of the list)
will be removed during the next update.

To reduce the need for repeated downloads, you can increase the number of stored versions by
manually setting the max_tools value to the desired limit.

> ### Managed updates initiation

Reviewer: ```suggestion
### Managed updates version checks
```

Client tools make a request to the Teleport Proxy Service discovery endpoint `/v1/webapi/find` during every login
in order to retrieve the required version of Teleport client tools and check the update mode for the target cluster.

To enable version checks for `tsh ssh` or `tsh proxy ssh` commands in case a login is not required,
set the `TELEPORT_TOOLS_CHECK_UPDATE=t` environment variable to initiate this check. For example:

```code
$ TELEPORT_TOOLS_CHECK_UPDATE=t tsh ssh user@host
Update progress: [▒▒▒▒▒▒▒▒▒▒] (Ctrl-C to cancel update)
user@host ~ #
```

Managed update local checks are triggered during the execution of any command. If there is an active
session with a cluster and the cluster advertises a required client tools version,
both `tsh` and `tctl` will be re-executed using that version. If the version is not already present
in the tools directory, it will be downloaded, extracted, and executed automatically.

If there is no active session, but a versioned binary exists in the tools directory
(e.g., `~/.tsh/bin/tsh` or `~/.tsh/bin/tctl`), the corresponding tool will be re-executed with that version
(for backward compatibility with older versions of managed updates).

> ## Configuring client tool automatic updates

Reviewer 2: ```suggestion
## Configuring client tool managed updates
```

> ### Using environment variables

Reviewer: ```suggestion
### On a per-client basis using environment variables
```

Values:
- `X.Y.Z`: Use a specific version.
- `off`: Disable updates.

An environment variable `TELEPORT_TOOLS_VERSION` can be used as an emergency workaround for a known issue,

> pinning to a specific version in CI/CD, for debugging, or for manual updates.

Reviewer: ```suggestion
The `TELEPORT_TOOLS_VERSION` environment variable can be used for pinning a specific version, for debugging, or for manual updates.

Possible values are:
- `X.Y.Z`: Use a specific version.
- `off`: Disable updates.
```

Update progress: [▒▒▒▒▒▒▒▒▒▒] (Ctrl-C to cancel update)
Teleport v18.1.0 git:v18.1.0-0-g8cdb161 go1.24.5

# If the environment variable is not set, the currently installed version is used.
$ tctl version
Teleport v17.6.0 git:v17.6.0-0-g4c3b13b go1.23.11
````

To use the version specified by the environment variable, it must always be set.
If the variable is set during the login command, we prioritize this version over the one
provided by the cluster configuration. Even if the cluster indicates that client tools
managed updates are disabled, setting the environment variable will force them to be enabled.
Any tsh or tctl command executed with an active profile for this cluster will be re-executed
using the recorded version.

```bash
$ TELEPORT_TOOLS_VERSION=18.1.0 tsh login --proxy proxy.example.com
Update progress: [▒▒▒▒▒▒▒▒▒▒] (Ctrl-C to cancel update)
...

# The session is still active, and the profile at ~/.tsh/current-profile is set to proxy.example.com.
$ tctl version
Teleport v18.1.0 git:v18.1.0-0-g8cdb161 go1.24.5

$ cat ~/.tsh/bin/.config.json
{
  "version": "v1",
  "configs": {
    "proxy.example.com": {"version": "18.1.0","disabled": false}
  }
}
````

> ### Using `tctl`

Reviewer: ```suggestion
### Cluster-wide using `tctl`
```

Also I think this section should be put before the per-client one, as this is what we want people to do.

> To enable or disable client tools automatic updates in the cluster, use the following command:

Reviewer 2: ```suggestion
To enable or disable client tools managed updates in the cluster, use the following command:
```

```code
$ tctl autoupdate client-tools enable
client tools auto update mode has been changed

$ tctl autoupdate client-tools disable
client tools auto update mode has been changed
```

To set or remove the target version for automatic updates for all client tools:

```code
$ tctl autoupdate client-tools target X.Y.Z
client tools auto update target version has been set

$ tctl autoupdate client-tools target --clear
client tools auto update target version has been cleared
```

If the target version is cleared, the cluster version will be used automatically, eliminating the need for manual
updates each time the cluster version is upgraded.

The `status` command retrieves the target version and mode configured for the logged-in cluster.
To use an unauthenticated endpoint for this configuration, include the `--proxy` flag.

```code
$ tctl autoupdate client-tools status --format json
{
    "mode": "enabled",
    "target_version": "X.Y.Z"
}

$ tctl autoupdate client-tools status --proxy proxy.example.com --format json
{
    "mode": "enabled",
    "target_version": "X.Y.Z"
}
```

### Using resource definitions

> To enable client tools automatic updates in cluster, first create a file named `autoupdate_config.yaml` with the following content:

Reviewer 2: ```suggestion
To enable client tools managed updates in cluster, first create a file named `autoupdate_config.yaml` with the following content:
```

```yaml
kind: autoupdate_config
metadata:
  name: autoupdate-config
spec:
  tools:
    mode: enabled
```

And write resource data to the cluster `tctl create -f autoupdate_config.yaml`, after that any new `tsh` login must
check the target version and initiate downloading desired version to install in Teleport home folder.

The next resource is responsible for setting target version `autoupdate_version.yaml`.

```yaml
kind: autoupdate_version
metadata:
  name: autoupdate-version
spec:
  tools:
    target_version: X.Y.Z
```

Create the resource using `tctl create -f autoupdate_version.yaml`.
If the `autoupdate_version` resource hasn't been created yet, the cluster version will be used as the default target version.

<Admonition type="note">

>     - For self-hosted clusters, automatic updates are disabled by default but can be enabled.

Reviewer 2: ```suggestion
    - For self-hosted clusters, managed updates are disabled by default but can be enabled.
```

    - Cloud clusters are automatically enrolled in updates, managed by the Teleport Cloud team.
    - For clusters with multiple root versions, use self-managed updates to avoid frequent version switching.
</Admonition>

## Determining a client tool version

To determine the version required to operate with the cluster, during the login process, `tsh` queries from the
unauthenticated proxy discovery `/v1/webapi/find` endpoint. If `.auto_update.tools_auto_update` is enabled, the
client tools must initiate the installation of the version specified in `.auto_update.tools_version`.

For manual updates, when scheduling updates at specific times or using custom CDN mirrors or with self-build packages,

> you can disable auto-update via configuration. In this case, you can monitor the tool's version separately

Reviewer 2: ```suggestion
you can disable managed updates via configuration. In this case, you can monitor the tool's version separately
```

To reduce the need for repeated downloads, you can increase the number of stored versions by
manually setting the max_tools value to the desired limit.

### Managed updates initiation

> Client tools make a request to the proxy discovery endpoint `/v1/webapi/find` during every login

Reviewer 3: ```suggestion
Client tools make a request to the Teleport Proxy Service discovery endpoint `/v1/webapi/find` during every login
```
For consistent naming in the docs

> in order to retrieve the required client tools version and check the update mode for the target cluster.

Reviewer 3: ```suggestion
in order to retrieve the required version of Teleport client tools and check the update mode for the target cluster.
```
Style suggestion

> To enable version checks for `tsh ssh` or `tsh proxy ssh` commands in case when login is not required,

Reviewer 3: ```suggestion
To enable version checks for `tsh ssh` or `tsh proxy ssh` commands in case a login is not required,
```
Grammar suggestion

set the `TELEPORT_TOOLS_CHECK_UPDATE=t` environment variable to initiate this check. For example:

> ```bash

Reviewer 3: ```suggestion
```code
```
So the command snippet renders as expected on the docs site
